File: CVcodes/contours.py
import cv2
import sys
import numpy as np

img = cv2.imread(sys.argv[1])
gray = cv2.cvtColor(img, cv2.cv.CV_BGR2GRAY)
blur = cv2.GaussianBlur(gray, (3, 3), 0)

thresh=100

canny_img = cv2.Canny(blur, threshold1=thresh, 
            threshold2=thresh*2, apertureSize=3)
contours, hierarchy = cv2.findContours(np.copy(canny_img), cv2.RETR_EXTERNAL, 
                    cv2.CHAIN_APPROX_NONE)
drawing = np.zeros(img.shape)
for i in range(len(contours)):
    cv2.drawContours(drawing, contours, i, (0, 0, 255), 1, 8, hierarchy=hierarchy)

cv2.imshow("image", img)
cv2.imshow("contours", drawing)

cv2.waitKey(0)

File: CVcodes/contours.cpp
#include "opencv2/highgui/highgui.hpp"
#include "opencv2/imgproc/imgproc.hpp"
#include <stdio.h>

using namespace cv;
using namespace std;

int main(int argc, char* argv[])
{
    Mat img, gray, blur, canny_img, drawing;
    img = imread(argv[1]);
    cvtColor(img, gray, CV_BGR2GRAY);
    GaussianBlur(gray, blur, Size(3, 3), 0);

    int thresh = 100;
    Canny(blur, canny_img, thresh, thresh*2, 3);

    vector<vector<Point> > contours;
    vector<Vec4i> hierarchy;

    drawing = Mat::zeros(canny_img.size(), CV_8UC3);
    findContours(canny_img, contours, hierarchy, CV_RETR_TREE, CV_CHAIN_APPROX_SIMPLE, Point(0,0));
    printf("%ld contours\n", contours.size());
    for (int i=0; i<contours.size(); i++)
    {
        drawContours(drawing, contours, i, Scalar(0, 0, 255), 2, 8, hierarchy, 0, Point());
    }
    imshow("image", img);
    imshow("contours", drawing);
    waitKey(0);

    return(0);
}